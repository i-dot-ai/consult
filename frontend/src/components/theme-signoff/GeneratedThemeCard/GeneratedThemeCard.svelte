<script lang="ts">
  import clsx from "clsx";

  import { slide, fly } from "svelte/transition";

  import type { GeneratedTheme, SelectedTheme } from "../../../global/types";
  import { createFetchStore } from "../../../global/stores";

  import Panel from "../../dashboard/Panel/Panel.svelte";
  import Button from "../../inputs/Button/Button.svelte";
  import MaterialIcon from "../../MaterialIcon.svelte";
  import ChevronRight from "../../svg/material/ChevronRight.svelte";
  import GeneratedThemeCard from "./GeneratedThemeCard.svelte";
  import Tag from "../../Tag/Tag.svelte";
  import AnswersList from "../AnswersList/AnswersList.svelte";
  import Visibility from "../../svg/material/Visibility.svelte";

  export interface Props {
    consultationId: string;
    selectedThemes: SelectedTheme[];
    theme: GeneratedTheme;
    level?: number;
    leftPadding?: number;
    forceExpand?: boolean;
    setForceExpand?: (newValue: boolean) => void;
    handleSelect: (theme: GeneratedTheme) => void;
    maxAnswers?: number;
  }
  let {
    consultationId,
    selectedThemes = [],
    theme,
    level = 0,
    leftPadding = 2,
    forceExpand = false,
    setForceExpand = () => {},
    handleSelect = () => {},
    maxAnswers = 10,
  }: Props = $props();

  let expanded = $state(true);
  let showAnswers = $state(false);
  let answersRequested = $state(false);

  let {
    load: loadAnswers,
    loading: isAnswersLoading,
    data: answersData,
    error: answersError,
  } = createFetchStore();

  let disabled = $derived(
    Boolean(
      selectedThemes.find((selectedTheme) => selectedTheme.id === theme.id),
    ),
  );

  $effect(() => {
    if (forceExpand) {
      expanded = true;
    }
  });

  const shouldShowAnswers = () => {
    return showAnswers && !disabled;
  };
</script>

<div
  transition:slide
  style="margin-left: {level * leftPadding}rem;"
  class={clsx(["generated-theme-card"])}
>
  <Panel border={true}>
    <article class="flex">
      <div
        class={clsx([
          "transition-all",
          "duration-300",
          shouldShowAnswers() ? "w-1/3" : "w-full",
        ])}
      >
        <div class="flex items-center justify-between mb-2">
          <div class="flex items-center gap-2">
            {#if theme.children?.length! > 0}
              <Button
                variant="ghost"
                handleClick={() => {
                  expanded = !expanded;
                  setForceExpand(false);
                }}
              >
                <div
                  class={clsx([
                    "transition-transform",
                    expanded && "rotate-90",
                  ])}
                >
                  <MaterialIcon size="1.2rem" color="fill-emerald-700">
                    <ChevronRight />
                  </MaterialIcon>
                </div>
              </Button>
            {/if}

            <h3>{theme.name}</h3>

            <div class={clsx([
              disabled && "grayscale",
            ])}>
              <Tag variant="success">
                Level {level + 1}
              </Tag>
            </div>
          </div>
        </div>

        <p class="text-neutral-500 text-sm">
          {theme.description}
        </p>

        {#if !disabled}
          <footer class="mt-4 flex items-center gap-2">
            <Button
              variant="approve"
              size="sm"
              handleClick={() => {
                disabled = !disabled;
                handleSelect(theme);
              }}
            >
              Select
            </Button>
            <Button
              size="sm"
              handleClick={() => {
                if (!$answersData) {
                  loadAnswers(`
                    /api/consultations/${consultationId}/responses/?searchValue=${theme.name}&searchMode=semantic
                  `);
                }
                showAnswers = !showAnswers;
                answersRequested = true;
              }}
              disabled={$isAnswersLoading && answersRequested}
            >
              <div class="text-emerald-700 flex items-center gap-1">
                <MaterialIcon color="fill-emerald-700">
                  <Visibility />
                </MaterialIcon>
                {showAnswers ? "Hide" : "Representative"} Responses
              </div>
            </Button>
          </footer>
        {/if}
      </div>

      {#if shouldShowAnswers()}
        <aside
          transition:fly={{ x: 300 }}
          class="border-l border-neutral-200 ml-4 pl-4"
        >
          <AnswersList
            variant="generated"
            title="Representative Responses"
            answers={
              $answersData?.all_respondents
                .slice(0, maxAnswers)
                .map(answer => answer.free_text_answer_text)
            || []}
          />
        </aside>
      {:else}
        <div class="p2"></div>
      {/if}
    </article>
  </Panel>

  {#if expanded}
    <div transition:slide class="pt-4">
      {#each theme.children as childTheme (childTheme.id)}
        <GeneratedThemeCard
          {consultationId}
          {selectedThemes}
          theme={childTheme}
          level={level + 1}
          {forceExpand}
          {setForceExpand}
          {handleSelect}
        />
      {/each}
    </div>
  {:else}
    <div class="pb-4"></div>
  {/if}
</div>

<style>
  .generated-theme-card :global(div[data-testid="panel-component"]) {
    margin-block: 0;
  }
</style>
