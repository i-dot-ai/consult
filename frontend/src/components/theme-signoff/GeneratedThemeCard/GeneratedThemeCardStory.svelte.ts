import GeneratedThemeCard from "./GeneratedThemeCard.svelte";

let selectedThemes = $state([]);
let theme = $state({
  id: "theme-id",
  name: "Theme Name",
  description: "Theme description",
});
let answers = $state(["Answer 1", "Answer 2"]);
let level = $state(0);
let leftPadding = $state(1);
let handleSelect = (theme) =>
  alert(`Select theme event triggered with: ${theme}`);

export default {
  name: "GeneratedThemeCard",
  component: GeneratedThemeCard,
  category: "Theme Signoff",
  props: [
    {
      name: "selectedThemes",
      value: selectedThemes,
      type: "json",
    },
    {
      name: "theme",
      value: theme,
      type: "json",
    },
    {
      name: "answers",
      value: answers,
      type: "json",
    },
    {
      name: "level",
      value: level,
      type: "number",
    },
    {
      name: "leftPadding",
      value: leftPadding,
      type: "number",
    },
    {
      name: "handleSelect",
      value: handleSelect,
      type: "func",
      schema: `(theme: GeneratedTheme) => void`,
    },
  ],
  stories: [],
};
