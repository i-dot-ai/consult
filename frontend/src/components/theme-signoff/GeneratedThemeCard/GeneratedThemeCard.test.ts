import { afterEach, beforeEach, describe, expect, it } from "vitest";
import { render, cleanup } from "@testing-library/svelte";

import GeneratedThemeCard, { type Props } from "./GeneratedThemeCard.svelte";

let testData: Props;

describe("GeneratedThemeCard", () => {
  beforeEach(() => {
    testData = {
      selectedThemes: [],
      theme: {
        id: "test-theme",
        name: "Test Theme",
        description: "This is a test theme",
      },
      answers: ["Answer 1", "Answer 2"],
      handleSelect: () => {},
    };
  })

  afterEach(() => cleanup());

  it("should render", async () => {
    const { container, getByText, getAllByText, queryByText } = render(GeneratedThemeCard, {
      ...testData,
    });

    expect(getByText(testData.theme.name));
    expect(getByText(testData.theme.description));

    // Answers hidden initially
    testData.answers?.forEach(answer => {
      expect(queryByText(answer)).toBeNull();
    })

    expect(container).toMatchSnapshot();
  });

  it("should render recursively", async () => {
    const CHILD_THEME = {
      id: "child-theme",
      name: "Child Theme",
      description: "This is a child theme",
    }
    const { container, getByText, getAllByText, queryByText } = render(GeneratedThemeCard, {
      ...testData,
      theme: {
        ...testData.theme,
        children: [
          CHILD_THEME,
        ]
      }
    });

    expect(getByText(CHILD_THEME.name));
    expect(getByText(CHILD_THEME.description));
  });
});
