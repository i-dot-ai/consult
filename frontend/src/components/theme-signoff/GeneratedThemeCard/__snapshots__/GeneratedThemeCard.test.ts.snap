// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GeneratedThemeCard > should render 1`] = `
<div>
  <div
    class="generated-theme-card svelte-g4v74i"
    style="margin-left: 0rem;"
  >
    <div
      class="h-full my-4 p-4 rounded-2xl border border-gray-200"
      data-testid="panel-component"
    >
      <article
        class="flex"
      >
        <div
          class="transition-all duration-300 w-full svelte-g4v74i"
        >
          <div
            class="flex items-center justify-between mb-2"
          >
            <div
              class="flex items-center gap-2"
            >
              <!---->
               
              <h3>
                Test Theme
              </h3>
               
              <div
                class="flex justify-between items-center gap-2 rounded-md w-max py-0.5 px-2 text-xs text-emerald-700 bg-emerald-50 border border-emerald-400"
              >
                Level 1
                <!---->
              </div>
              <!---->
            </div>
          </div>
           
          <p
            class="text-neutral-500 text-sm"
          >
            This is a test theme
          </p>
           
          <footer
            class="mt-4 flex items-center gap-2"
          >
            <!---->
            <button
              aria-pressed="false"
              class="text-sm cursor-pointer rounded-md py-1 px-2 border bg-emerald-700 text-white border-gray-300 transition-colors duration-300 self-start inline-flex gap-1 items-center hover:bg-gray-100 hover:text-teal hover:border-teal svelte-1rjb6fa"
              data-variant="approve"
              role="button"
              tabindex="0"
            >
              <!---->
              Select
              
              
            </button>
            
            <!---->
             
            <!---->
            <button
              aria-pressed="false"
              class="text-sm cursor-pointer rounded-md py-1 px-2 border bg-white border-gray-300 transition-colors duration-300 self-start inline-flex gap-1 items-center hover:bg-gray-100 svelte-1rjb6fa"
              data-variant="default"
              role="button"
              tabindex="0"
            >
              <!---->
              <div
                class="text-emerald-700 flex items-center gap-1"
              >
                <svg
                  class="fill-emerald-700"
                  height="1rem"
                  role="img"
                  viewBox="0 -960 960 960"
                  width="1rem"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M480-320q75 0 127.5-52.5T660-500q0-75-52.5-127.5T480-680q-75 0-127.5 52.5T300-500q0 75 52.5 127.5T480-320Zm0-72q-45 0-76.5-31.5T372-500q0-45 31.5-76.5T480-608q45 0 76.5 31.5T588-500q0 45-31.5 76.5T480-392Zm0 192q-146 0-266-81.5T40-500q54-137 174-218.5T480-800q146 0 266 81.5T920-500q-54 137-174 218.5T480-200Zm0-300Zm0 220q113 0 207.5-59.5T832-500q-50-101-144.5-160.5T480-720q-113 0-207.5 59.5T128-500q50 101 144.5 160.5T480-280Z"
                    xmlns="http://www.w3.org/2000/svg"
                  />
                  <!---->
                </svg>
                <!---->
                 Representative Responses
              </div>
              
              
            </button>
            
            <!---->
          </footer>
          <!---->
        </div>
         
        <div
          class="p2"
        />
        <!---->
      </article>
      <!---->
    </div>
    <!---->
     
    <div
      class="pt-4"
    >
      
    </div>
    <!---->
  </div>
  
</div>
`;
